include/ringbuffer_utils.hpp

```diff
@@ -152,6 +153,151 @@ class PerfRingBufferReader {
   uint64_t _head;
 };
 
+struct MPSCRingBufferHeader {
+  uint64_t size;
+  static constexpr uint64_t k_discard_bit = 1UL << 62;
+  static constexpr uint64_t k_busy_bit = 1UL << 63;
+
+  static bool is_busy(uint64_t size) { return size & k_busy_bit; }
+  static bool is_discarded(uint64_t size) { return size & k_discard_bit; }
+};
+
+class MPSCRingBufferWriter {
+public:
+  explicit MPSCRingBufferWriter(RingBuffer &rb) : _rb(rb) {
+    assert(rb.type == RingBufferType::kMPSCRingBuffer);
+    _tail = __atomic_load_n(_rb.reader_pos, __ATOMIC_ACQUIRE);
```

Comment:
Minor: I think this can be slightly simplified:

```suggestion
    update_tail();
```